Sources/GRPC/ClientCalls/Call.swift

```diff
@@ -0,0 +1,279 @@
+/// traverse the pipeline having been received from the network.
+///
+/// This object is a lower-level API than the equivalent wrapped calls (such as `UnaryCall` and
+/// `BidirectionalStreamingCall`). The caller is therefore required to do more in order to use this
+/// object correctly. Callers must call `invoke(_:)` to start the call and ensure that the correct
+/// number of request parts are sent in the correct order (exactly one `metadata`, followed
+/// by at most one `message` for unary and server streaming calls, and any number of `message` parts
+/// for client streaming and bidirectional streaming calls. All call types must terminate their
+/// request stream by sending one `end` message.
+///
+/// Callers are not able to create `Call` objects directly, rather they must be created via an
+/// object conforming to `GRPCChannel` such as `ClientConnection`.
+public class Call<Request, Response> {
+  private enum State {
+    /// Idle, waiting to be invoked.
+    case idle(ClientTransportFactory<Request, Response>)
+
+    /// Invoked, we have a transport on which to send requests. The transport may be closed if the
+    /// RPC has already completed.
+    case invoked(ClientTransport<Request, Response>)
+  }
+
+  /// The current state of the call.
+  private var state: State
+
+  /// User provided interceptors for the call.
+  private let interceptors: [ClientInterceptor<Request, Response>]
+
+  /// Whether compression is enabled on the call.
+  private var isCompressionEnabled: Bool {
+    return self.options.messageEncoding.enabledForRequests
+  }
+
+  /// The `EventLoop` the call is being invoked on.
+  public let eventLoop: EventLoop
+
+  /// The path of the RPC, usually generated from a service definition, e.g. "/echo.Echo/Get".
+  public let path: String
+
+  /// The type of the RPC, e.g. unary, bidirectional streaming.
+  public let type: GRPCCallType
+
+  /// Options used to invoke the call.
+  public let options: CallOptions
+
+  // Calls can't be constructed directly: users must make them using a `GRPCChannel`.
+  internal init(
+    path: String,
+    type: GRPCCallType,
+    eventLoop: EventLoop,
+    options: CallOptions,
+    interceptors: [ClientInterceptor<Request, Response>],
+    transportFactory: ClientTransportFactory<Request, Response>
+  ) {
+    self.path = path
+    self.type = type
+    self.options = options
+    self.state = .idle(transportFactory)
+    self.eventLoop = eventLoop
+    self.interceptors = interceptors
+  }
+
+  /// Starts the call and provides a callback which is invoked on every response part received from
+  /// the server.
+  ///
+  /// This must be called prior to `send(_:promise:)` or `cancel(promise:)`.
+  ///
+  /// - Parameter onResponsePart: A callback which is invoked on every response part.
+  /// - Important: This function should only be called once. Subsequent calls will be ignored.
+  public func invoke(_ onResponsePart: @escaping (ClientResponsePart<Response>) -> Void) {
```

Comment:
There are lots of generic functions here: should we consider making them inlinable?

Reply:
Yes, good thinking.